docs/source/components.rst

```diff
@@ -0,0 +1,332 @@
+from ``Form.__init__()`` if you overrode that) is done in a meta class. This
+meta class could be documented and exposed as public API to be able to easily
+subclass it. The disadvantage with this approach is that you can't easily pass
+any custom arguments to class creation (i.e., the meta class' ``__new__()`` and
+``__init__()``). However those are often needed with ``request.user`` being one
+of the examples.
+
+Another solution is to use a factory function that can take a variable number of
+arguments and returns a new class similar to how ``formset_factory()`` currently
+works.
+
+
+Existing Data
+=============
+
+All data that is provided by the system in one way or another and was not input
+by the user is called "existing data". This includes but is not limited to field
+defaults and initial data. This step describes the process of collecting the
+existing data from all sources. The goal is to have an API that makes it
+possible to easily add more sources of existing data and to customize it on a
+per-request basis, e.g. "fetch the existing data from source A if user is admin,
+otherwise fetch it from source B".
+
+How it is currently done
+------------------------
+
+Form field defaults can be specified with the ``initial`` argument to
+``Field()``:
+
+.. code-block:: python
+
+    title = CharField(
+        max_length=200,
+        initial='Enter your awesome title here!'
+    )
+
+
+It is also possible to pass a dictionary of initial data to ``Form.__init__()``:
+
+.. code-block:: python
+
+    my_initial_data = {'title': 'Enter your awesome title here!'}
+    form = BlogPostForm(initial=my_initial_data)
+
+
+How it could be done
+--------------------
+
+See "How it could be done" on the section "New Data".
+
+
+New Data
+========
+
+All data that is provided by external sources (e.g. an HTML form submitted by a
+user or POSTed JSON data) is called "new data". For the moment we'll call them
+"Input Serializers". An Input Serializer must follow a documented API that
+allows data of all kinds and shapes to be converted to a universal data
+structure. That data structure then serves a data source for other components.
+
+How it is currently done
+------------------------
+
+In Django Forms user data is passed to a form instance via the ``data`` and
+``files`` argument:
+
+.. code-block:: python
+
+    form = BlogPostForm(data=request.POST, files=request.FILES)
+
+
+The conversion to a "universal" data structure then happens in
+``Widget.value_from_datadict()``. This method handles all potential oddities of
+an HTML form, like multiple input fields with the same name or different input
+fields belonging to one logical field.
+
+DRF uses the ``data`` argument to a Serializer:
+
+.. code-block:: python
+
+    serializer = BlogPostSerializer(data=json_data)
+
+
+How it could be done
+--------------------
+
+At first the universal data structure has to be defined. It should not be
```

Comment:
It's worth looking at `Field.to_python` - this is currently the start of [the validation chain](https://docs.djangoproject.com/en/1.9/ref/forms/validation/) and it handles stuff like ensuring a `FloatField` value *is* a `float`.